From 8412c495bc9d4b7ce072728b77c48fba512a010a Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Mon, 5 Jun 2023 13:49:10 +0200
Subject: [PATCH] pinctrl: stm32: pass bank_ioport_nr inside the interrupt
 specifier

To decouple pinctrl from the interrupt controller, pass the bank
number in bits [23:16] of the second interrupt specifier.
This will be used by the EXTI interrupt controller to program the
mux that selects the gpio bank.

Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I369827e80243c248e2a67ad4aa7d79a55a87b06d
---
 drivers/pinctrl/stm32/pinctrl-stm32.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

--- a/drivers/pinctrl/stm32/pinctrl-stm32.c
+++ b/drivers/pinctrl/stm32/pinctrl-stm32.c
@@ -90,8 +90,11 @@
 #define STM32_GPIO_BKP_DELAY_SHIFT	16
 #define STM32_GPIO_BKP_DELAY_MASK	GENMASK(19, 16)
 
-#define STM32_GPIO_PINS_PER_BANK 16
-#define STM32_GPIO_IRQ_LINE	 16
+#define STM32_GPIO_PINS_PER_BANK	16
+#define STM32_GPIO_IRQ_LINE		16
+
+/* to pass bank_ioport_nr to EXTI in struct irq_fwspec::param[1] */
+#define STM32_GPIO_BANK_MASK		GENMASK(23, 16)
 
 #define SYSCFG_IRQMUX_MASK GENMASK(3, 0)
 
@@ -655,7 +658,8 @@ static int stm32_gpio_domain_alloc(struc
 	parent_fwspec.fwnode = d->parent->fwnode;
 	parent_fwspec.param_count = 2;
 	parent_fwspec.param[0] = fwspec->param[0];
-	parent_fwspec.param[1] = fwspec->param[1];
+	parent_fwspec.param[1] = (fwspec->param[1] & IRQ_TYPE_SENSE_MASK) |
+				 FIELD_PREP(STM32_GPIO_BANK_MASK, bank->bank_ioport_nr);
 
 	irq_domain_set_hwirq_and_chip(d, virq, hwirq, &stm32_gpio_irq_chip,
 				      bank);
